components/gfx/paint_context.rs

```diff
@@ -466,6 +466,167 @@ impl<'a> PaintContext<'a> {
         }
     }
 
+    fn is_zero_radius(radius: &Size2D<AzFloat>) -> bool {
+        radius.width <= 0. || radius.height <= 0.
+    }
+
+    // Adapted from gecko:gfx/2d/PathHelpers.h:EllipseToBezier
+    fn ellipse_to_bezier(path_builder: &mut PathBuilder,
+                         origin: Point2D<AzFloat>,
+                         radius: Size2D<AzFloat>,
+                         start_angle: f32,
+                         end_angle: f32) {
+        if PaintContext::is_zero_radius(&radius) {
+            return;
+        }
+
+        // Calculate kappa constant for partial curve. The sign of angle in the
+        // tangent will actually ensure this is negative for a counter clockwise
+        // sweep, so changing signs later isn't needed.
+        let kappa_factor: f32 = (4.0f32 / 3.0f32) * ((end_angle - start_angle) / 4.).tan();
+        let kappa_x: f32 = kappa_factor * radius.width;
+        let kappa_y: f32 = kappa_factor * radius.height;
+
+        // We guarantee here the current point is the start point of the next
+        // curve segment.
+        let start_point = Point2D::new(origin.x + start_angle.cos() * radius.width,
+                                       origin.y + start_angle.sin() * radius.height);
+
+        path_builder.line_to(start_point);
+        let end_point = Point2D::new(origin.x + end_angle.cos() * radius.width,
+                                     origin.y + end_angle.sin() * radius.height);
+
+        let tangent_start = Point2D::new(-start_angle.sin(), start_angle.cos());
+
+        let cp1 = Point2D::new(start_point.x + tangent_start.x * kappa_x,
+                               start_point.y + tangent_start.y * kappa_y);
+
+        let rev_tangent_end = Point2D::new(end_angle.sin(), -end_angle.cos());
+
+        let cp2 = Point2D::new(end_point.x + rev_tangent_end.x * kappa_x,
+                               end_point.y + rev_tangent_end.y * kappa_y);
+
+        path_builder.bezier_curve_to(&cp1, &cp2, &end_point);
+    }
+
+    #[allow(non_snake_case)]
+    fn border_bounds(bounds: &Rect<f32>,
+                     border: &SideOffsets2D<f32>) -> ((Point2D<f32>, Point2D<f32>),
+                                                      (Point2D<f32>, Point2D<f32>),
+                                                      (Point2D<f32>, Point2D<f32>),
+                                                      (Point2D<f32>, Point2D<f32>)) {
```

Comment:
I'd think, I'd have to add duplicated code in the 

    let ((box_TL, inner_TL), (box_TR, inner_TR), ...) = PaintContext::border_bounds(...) 

assignments to calculate `box_TL, inner_TL, box_TR, inner_TR`, if I changed it to two Rects.